x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/build_alert.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SearchTypes } from '../../../../../../common/detection_engine/types';
+import { RulesSchema } from '../../../../../../common/detection_engine/schemas/response/rules_schema';
+import { isEventTypeSignal } from '../../../signals/build_event_type_signal';
+import { Ancestor, BaseSignalHit, ThresholdResult } from '../../../signals/types';
+import { getValidDateFromDoc } from '../../../signals/utils';
+import { RACAlertSignalWithRule } from '../../types';
+
+/**
+ * Takes a parent signal or event document and extracts the information needed for the corresponding entry in the child
+ * signal's `signal.parents` array.
+ * @param doc The parent signal or event
+ */
+export const buildParent = (doc: BaseSignalHit): Ancestor => {
+  if (doc._source?.signal != null) {
+    return {
+      rule: doc._source?.signal.rule.id,
+      id: doc._id,
+      type: 'signal',
+      index: doc._index,
+      // We first look for signal.depth and use that if it exists. If it doesn't exist, this should be a pre-7.10 signal
+      // and should have signal.parent.depth instead. signal.parent.depth in this case is treated as equivalent to signal.depth.
+      depth: doc._source?.signal.depth ?? doc._source?.signal.parent?.depth ?? 1,
+    };
+  } else {
+    return {
+      id: doc._id,
+      type: 'event',
+      index: doc._index,
+      depth: 0,
+    };
+  }
+};
+
+/**
+ * Takes a parent signal or event document with N ancestors and adds the parent document to the ancestry array,
+ * creating an array of N+1 ancestors.
+ * @param doc The parent signal/event for which to extend the ancestry.
+ */
+export const buildAncestors = (doc: BaseSignalHit): Ancestor[] => {
+  const newAncestor = buildParent(doc);
+  const existingAncestors = doc._source?.signal?.ancestors;
+  if (existingAncestors != null) {
+    return [...existingAncestors, newAncestor];
+  } else {
+    return [newAncestor];
+  }
+};
+
+/**
+ * This removes any signal name clashes such as if a source index has
+ * "signal" but is not a signal object we put onto the object. If this
+ * is our "signal object" then we don't want to remove it.
+ * @param doc The source index doc to a signal.
+ */
+export const removeClashes = (doc: BaseSignalHit): BaseSignalHit => {
+  // @ts-expect-error @elastic/elasticsearch _source is optional
+  const { signal, ...noSignal } = doc._source;
```

Comment:
I would propose checking if the `_source` field is in `doc` instead of using `@ts-expect-error`. 

In my [PR](https://github.com/elastic/kibana/pull/103463/files#diff-b9f1366a9a19b657da0671cc4224d1fb6735b7d243985ca4e1e2f625277bb9b2), I've added a helper function to perform this kind of check; however, it is not yet merged. Here is its source:

```ts
export class InvariantError extends Error {
  name = 'Invariant violation';
}

/**
 * Asserts that the provided condition is always true
 * and throws an invariant violation error otherwise
 *
 * @param condition Condition to assert
 * @param message Error message to throw if the condition is falsy
 */
export function invariant(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new InvariantError(message);
  }
}
```

It could be used like that:

```ts
  invariant(doc_source, '_source field not found');
  const { signal, ...noSignal } = doc._source;
```

With it, we 
- don't need to use `@ts-expect-error` as Typescript narrows down the asserted type for us
- get a more readable error at runtime if something goes wrong: `Invariant violation: _source field not found` vs. `Cannot destructure property 'signal' of 'undefined' or 'null'`. Which would be much easier to debug in the case of SDH or something.

Reply:
Great suggestion, added!